api/src/Hanabi/Types.hs

```diff
@@ -119,42 +123,35 @@ redact pid = map (redactCard pid)
                (Just $ view cardRank card)
                (Just $ view cardColor card)
 
-applyWithInvalid :: Choice -> Game -> Either String Game
-applyWithInvalid (Choice pid playerChoice) state = do
-  unless (pid == view gameCurrentPlayer state) $
-    Left $ "cannot make a choice when not your turn"
-  
-  runApp state $
-    case playerChoice of
-      ChoicePlayCard cid -> applyPlay cid
-
-      ChoiceDiscardCard cid        -> applyDiscard cid
-      ChoiceHintRank target rank   -> applyHintRank target rank
-      ChoiceHintColor target color -> applyHintColor target color
+type AppEff effs = Members '[ Error String, State Game ] effs
 
+runApp :: Game -> Eff '[ Error String, State Game ] a -> Either String Game
 runApp state m =
-  let (val, newGame) = runMonadStack m state in
+  let (val, newGame) = run . runState state . runError $ m in
 
   val >>= \_ -> Right newGame
 
-  where
-    runMonadStack m state = runIdentity $ runStateT (runExceptT m) state
+applyWithInvalid :: Choice -> Game -> Either String Game
+applyWithInvalid (Choice pid playerChoice) state = do
+  unless (pid == view gameCurrentPlayer state) $
+    Left $ "cannot make a choice when not your turn"
 
-type App a = (ExceptT String (StateT Game Identity)) a
+  runApp state $ case playerChoice of
+    ChoicePlayCard cid           -> applyPlay cid
+    ChoiceDiscardCard cid        -> applyDiscard cid
+    ChoiceHintRank target rank   -> applyHintRank target rank
+    ChoiceHintColor target color -> applyHintColor target color
 
-requireCard2 :: CardId -> App Card
-requireCard2 cid =
+requireCard :: AppEff effs => CardId -> Eff effs Card
+requireCard cid =
   get >>= note "Card does not exist" . view (gameCards . at cid)
 
-requireCard :: CardId -> Game -> Either String Card
-requireCard cid = note "Card does not exist" . view (gameCards . at cid)
-
-applyPlay :: CardId -> App ()
+applyPlay :: AppEff effs => CardId -> Eff effs ()
 applyPlay cid = do
-  chosenCard <- requireCard2 cid
+  chosenCard <- requireCard cid
 
   maxRankOnTable <-
-    maximum
+    gets $ maximum
```

Comment:
This actually isn't relevant to `freer-simple`, I just remembered about `gets` and figured was more idiomatic than `<$>`.